### PR TITLE
Always sign request by adding Header fields

### PR DESCRIPTION
**Pull request by Ro-M**
- Always sign requests by adding header fields instead of signing the url
- Added a test to check that a signed header is used

### CODE REVIEW

**Comment by Ro-M** [comment]
relates to issue #266

File: Tests/AWSSDKSwiftCoreTests/AWSClientTests.swift [hunk 1]
```diff
@@ -406,6 +406,35 @@ class AWSClientTests: XCTestCase {
             XCTFail(error.localizedDescription)
         }
     }
+    
+    func testSignedClient() {
+        let input = E()
+        let client = AWSClient(
```
Comment (adam-fowler) [inline Tests/AWSSDKSwiftCoreTests/AWSClientTests.swift:412]: I recently added a helper function `createAWSClient()` which fills out all the parameters with default values. If you could use that instead that'd be great. Unless you want a specfic parameter set you can call `createAWSClient()` with nothing set. Means we have less client creation code filling up the tests.

File: Tests/AWSSDKSwiftCoreTests/AWSClientTests.swift [hunk 2]
```diff
@@ -406,6 +406,35 @@ class AWSClientTests: XCTestCase {
...
+            region: .useast1,
+            service: "s3",
+            serviceProtocol: .restxml,
+            apiVersion: "2013-12-02",
+            middlewares: [],
+            httpClientProvider: .createNew
+        )
+
+        do {
+            let awsRequest = try client.createAWSRequest(
```
Comment (adam-fowler) [inline Tests/AWSSDKSwiftCoreTests/AWSClientTests.swift:424]: Can you try this with different requests, maybe a GET request as well, seeing as that was originally the requests that used SignedURL.

Comment (Ro-M) [reply Tests/AWSSDKSwiftCoreTests/AWSClientTests.swift:424]: good point. would you like separate tests for that or is it ok if I test multiple requests in a single test?

**Comment by Ro-M** [comment]
on a side note: the `toHTTPRequestWithSignedURL` method on `AWSRequet` is now unused code and not public. Should I remove this method or should it remain for now?

File: Tests/AWSSDKSwiftCoreTests/AWSClientTests.swift (hunk 2 above)
Comment (adam-fowler) [reply Tests/AWSSDKSwiftCoreTests/AWSClientTests.swift]: It's ok to put both of those in the same test

**Comment by jonnymacs** [comment]
@adam-fowler @Ro-M hey guys, this is triggering some alerts for me. I believe signed headers don't work to auth head requests to s3 when running in a container that uses the IAM metadata credential service with a token. I think i lost a bit of my mind making this discovery.

please test that use case.

**Comment by adam-fowler** [comment]
> @adam-fowler @Ro-M hey guys, this is triggering some alerts for me. I believe signed headers don't work to auth head requests to s3 when running in a container that uses the IAM metadata credential service with a token. I think i lost a bit of my mind making this discovery.
> 
> please test that use case.

Seems a very odd failure. AWS do say everything works with headers, but we'll check before merging. @Ro-M you thought you had chosen an easy task as well.

**Comment by adam-fowler** [comment]
> on a side note: the `toHTTPRequestWithSignedURL` method on `AWSRequet` is now unused code and not public. Should I remove this method or should it remain for now?

Yes, if everything works

**Comment by adam-fowler** [comment]
@jonnymacs is that an ECS container? Or a docker image running on EC2?

**Comment by jonnymacs** [comment]
Ecs container.

**Comment by jonnymacs** [comment]
Yeah it was an odd and miserable failure to debug. 401 unauthorized, but
only when the token is required. It was about a year ago, there is a commit
that ensures HEAD request use the signedURL instead of headers that I made
to get it working.

File: Tests/AWSSDKSwiftCoreTests/AWSClientTests.swift [hunk 3]
```diff
@@ -406,6 +406,35 @@ class AWSClientTests: XCTestCase {
...
+                operation: "CopyObject",
+                path: "/",
+                httpMethod: "PUT",
+                input: input
+            )
+
+            let request: AWSHTTPRequest = awsRequest.createHTTPRequest(signer: try client.signer.wait())
+
+            XCTAssertNotNil(request.headers["Authorization"].first)
```
Comment (fabianfett) [inline Tests/AWSSDKSwiftCoreTests/AWSClientTests.swift]: if you just want to ensure no error is thrown, we (following @weissi's wisdom 😉) normally use a pattern like this: 

```swift
var request: AWSHTTPRequest?
XCTAssertNoThrow(request = awsRequest.createHTTPRequest(signer: try client.signer.wait()))
XCTAssertNotNil(request?.headers["Authorization"].first)
```

This saves us an indentation and ensures all lines in the test are always hit.

Comment (Ro-M) [reply Tests/AWSSDKSwiftCoreTests/AWSClientTests.swift]: sounds better this way. I just copied the way another test is written. I'll change that test too.

**Comment by fabianfett** [comment]
Hi @jonnymacs, just verified @Ro-M change works. On `ecs fargate` i tested with the following code:

```swift
import Vapor
import AWSS3

var env = try Environment.detect()
try LoggingSystem.bootstrap(from: &env)
let app = Application(env)
let s3 = S3(region: .eucentral1, httpClientProvider: .shared(app.http.client.shared))

struct Response: Content {
    var contentLength: Int64?
    var contentType: String?
}

app.get("hello") { req -> EventLoopFuture<Response> in
    
    s3.headBucket(.init(bucket: "my.bucket"))
        .flatMap { _ in
            // bucket exists
            s3.headObject(.init(bucket: "my.bucket", key: "03cd03760f29d253c5c689f691aa4343"))
        }
        .map { (output) -> (Response) in
            Response(contentLength: output.contentLength, contentType: output.contentType)
        }
        .flatMapErrorThrowing { (error) -> Response in
            throw error
        }
}

defer { app.shutdown() }
try app.run()
```

Does that void your concerns?

**Comment by jonnymacs** [comment]
@fabianfett is the client using the MetaDataCredentialProvider to obtain credentials?

**Comment by adam-fowler** [comment]
@jonnymacs I've seen this working on ECS as well, with the client using the MetaDataCredentialProvider for credentials. I'm really not sure what your issue was last year when you saw this. Maybe using a signed URL meant some other issue (which has since been fixed) was avoided. Although I can't think what it would be.

**Comment by jonnymacs** [comment]
Thanks guys. I retested my use case and indeed it's working with signed headers. I'll conclude it was fixed in the past year.

sorry for the fire drill!

**Comment by adam-fowler** [comment]
Sorry @Ro-M noticed you didnt get a chance to remove `toHTTPRequestWithSignedURL` before I merged. I'll do it

**Pull request merged**